src/vdo.rs

```diff
@@ -0,0 +1,482 @@
+                field: "dfp_vdo_version".into(),
+                value: dfp_vdo_version,
+                #[cfg(feature = "backtrace")]
+                backtrace: std::backtrace::Backtrace::capture(),
+            })?;
+
+        bit_reader.skip(2)?;
+
+        let host_capability = bit_reader.read(3)?;
+        let host_capability =
+            Pd3p2DfpVdoHostCapability::n(host_capability).ok_or_else(|| Error::ParseError {
+                field: "host_capability".into(),
+                value: host_capability,
+                #[cfg(feature = "backtrace")]
+                backtrace: std::backtrace::Backtrace::capture(),
+            })?;
+
+        bit_reader.skip(2)?;
+
+        let port_number = bit_reader.read(5)?;
+
+        Ok(Pd3p2DfpVdo {
+            dfp_vdo_version,
+            host_capability,
+            port_number,
+        })
+    }
+}
+
+/// The Discover Modes Command returns a list of zero to six VDOs, each of which
+/// describes a Mode.
+///
+/// See 6.4.4.2.4 Object Position in USB-PD
+pub const MAX_NUM_ALT_MODE: usize = 6;
+
+#[repr(C)]
+#[derive(Debug, Clone, PartialEq, Printf, Snprintf)]
+pub struct Pd3p2ProductVdo {
+    /// Product ID (assigned by the manufacturer)
+    product_id: u32,
+    /// Device release number.
+    device: BcdWrapper,
+}
+
+impl FromBytes for Pd3p2ProductVdo {
+    fn from_bytes(reader: &mut BitReader) -> Result<Self> {
+        let product_id = reader.read(16)?;
+        let device = reader.read(16)?;
+
+        Ok(Pd3p2ProductVdo {
+            product_id,
+            device: BcdWrapper(device),
+        })
+    }
+}
+
+/// Contains the XID assigned by USB-IF to the product before certification in
+/// binary format
+///
+/// See table 6.38 in the USB PD Specification for more information.
+#[repr(C)]
+#[derive(Debug, Clone, PartialEq, Printf, Snprintf)]
+pub struct Pd3p2CertStatVdo {
+    /// The XID assigned by USB-IF to the product before certification in binary
+    /// format.
+    pub xid: u32,
+}
+
+impl FromBytes for Pd3p2CertStatVdo {
+    fn from_bytes(reader: &mut BitReader) -> Result<Self> {
+        let xid = reader.read(32)?;
+
+        Ok(Pd3p2CertStatVdo { xid })
+    }
+}
+
+#[repr(C)]
+#[derive(Debug, Clone, PartialEq, Printf, Snprintf, N)]
+/// See USBPDB 6.4.4.3.1.4
+pub enum Pd3p2SopDfpProductType {
+    NotADfp,
+    PdUsbHub,
+    PdUsbHost,
+    PowerBrick,
+}
+
+#[repr(C)]
+#[derive(Debug, Clone, PartialEq, Printf, Snprintf, N)]
+pub enum Pd3p2SopUfpProductType {
+    NotAUfp,
+    PdUsbHub,
+    PdUsbPeripheral,
+    Psd,
+    NotACablePlugOrVPD,
+    PassiveCable,
+    ActiveCable,
+    VConnPoweredUsbDevice,
+}
+
+#[repr(C)]
+#[derive(Debug, Clone, PartialEq, Printf, Snprintf, N)]
+pub enum Pd3p2IdHeaderVdoConnectorType {
+    TypecReceptacle = 2,
+    TypecPlug = 3,
+}
+
+#[repr(C)]
+#[derive(Debug, Clone, PartialEq, Printf, Snprintf)]
+pub struct Pd3p2IdHeaderVdo {
+    /// Null-terminated vendor string.
```

Comment:
If you mean `String`, then it's because `String` is not `repr(C)`.

One of the requirements for this project is to have a C API. Hence, all structs exposed to the user are declared as `repr(C)` for interoperability. In particular, not only is String not repr(C), but it allocates, and allocations are a bit tricky with `cbindgen`.

For anything that allocates we have to either `Box::into_raw()` it, so that C gets a pointer that we can then reassemble with `Box::from_raw()` so that `drop` actually works, or something equivalent.

Also, C code will expect null-terminated strings, which won't be the case with `String`, whereas for `CString`, the same arguments above apply.

I think this is a good compromise, Rust users can convert that into `CStr` before using the field.

Reply:
Coming from a Rust perspective I wouldn't want compromises. I know you have been working on in-kernel Rust code and it makes sense that Rust engineers need to make compromises **in the kernel**, but this is offered as a normal Rust crate. As a user needing the features from this crate in my Rust code, I would just fork and "fix" the API to be less of a compromise. From Rust perspective there are no gains in the compromises.

I think you should least set the right expectations in the Readme.md and mention, that C API is considered first citizen for this crate even if that means having to make compromises in the Rust API.

Reply:
This is completely unrelated with Rust in the kernel, it's just that the clients have specifically asked for a C API so as to cater to C programs (that otherwise are currently linking against libtypec, i.e.: the C version of the library that we're trying to port into Rust)

I am open to other approaches, the problem is that in this particular case I haven't found any other one that can satisfy all requirements so far.

@sjoerdsimons your opinion here?

Reply:
Right now the Rust API is C-style in multiple places.  This make sense when C users have the highest priority in the project. It also means, that Rust users do take most of the compromises and need further wrapping to get native types and/or a more sensible API.

You can also do it the other way around. Make the API use all the fancy Rust stuff. Then for anything C cannot use, you add a wrapper function providing a C-style API. This effectively moves the wrapping from Rust users to C users, since the C-wrappers will not be used and thus eliminated by the compiler for Rust code. Thus Rust would be the highest priority and C becomes the second priority. Considering the code itself is written in Rust, that would be my default expectation.

Reply:
Let's just switch to `String` and Box it for C users. We can provide an extra C API function to extract a C string from the `String`.

The Rust users will enjoy a native api without paying any extra price.